### PR TITLE
Some fixes and 'allowed groups for bastion group' feature

### PR DESCRIPTION
**Pull request by Aleksey-Terzi**
**Fixes and tweaks:**

1) Fixed double messaging when player is right clicking in /bsi mode

2) Added new option commonSettings.cancelReinInBastionField
By default it is "false" and everything works as previously.
When it is "true" - then bastion will prevent "/ctr" action if player is not in bastion group and "/ctr" group is not allowed for bastion group

**New functionality:**

Introduced new feature "allowed groups for bastion group"

How it works:

1) Players could add groups to bastion groups

2) When some player haven't permissions on bastion group, but is member of "allowed group of this bastion" then he/she could place blocks using "/ctf allowed_group" or reinforce by using "/ctr allowed_group"even if commonSettings.cancelReinInBastionField = true

3) When such player (who is not in bastion group but in allowed group) is testing bastion field using /bsi - message will tell what group allowed to use

New permissions:

- BASTION_MANAGE_GROUPS - allows to modify "allowed group list"

New commands:

-  bsga:
      description: Adds allowed group to bastion group
      usage: /bsga <bastion group> <allowed group>
      permission: Bastion.normal
-  bsgd:
      description: Deletes allowed group from bastion group
      usage: /bsgd <bastion group> <allowed group>
      permission: Bastion.normal
-  bsgl:
      description: List allowed groups in bastion group
      usage: /bsga <bastion group>
      permission: Bastion.normal

### CODE REVIEW

**Comment by DevotedServant** [comment]
Can one of the admins verify this patch? @ProgrammerDan @Maxopoly

"ok to test" will build any commits made to this PR
"test this please" will build this PR once
"retest this please" will make another single build for this PR
"add to whitelist" will whitelist the creator of this PR and make jenkins automatically build any PRs made by him

**Comment by ProgrammerDan** [comment]
Need to fix the build for this, but all we need now is @Maxopoly to make a GUI for it!

**Comment by Maxopoly** [comment]
What would a Bastion GUI even do/show?

**Comment by ProgrammerDan** [comment]
In the context of this PR, it would show a management interface to add / remove "allowed" groups.

**Comment by Aleksey-Terzi** [comment]
> Need to fix the build for this

What do you mean about this?

About GUI - I thought to have some kind integration with /nl (if this possible at all and not so hard), something like:
1) Group's "page" will have new items/commands - add and manage "bastion allowed groups"
2) Add - works something like add new member to group
3) Manage  - allows to review list and remove existent groups

Also draw attention that player could add "allowed group" even if he/she is not member of this group.
In other words - I (as player) could allow to use any existent group in my bastion field.

File: src/main/java/isaac/bastion/BastionGroup.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+package isaac.bastion;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class BastionGroup {
+	private int groupId;
+	private Set<Integer> allowedGroupIds;
```
Comment (ProgrammerDan) [inline src/main/java/isaac/bastion/BastionGroup.java:12]: Maybe you've handled it elsewhere but doesn't this get us in trouble again? That whole "A group with one name may have many IDs pointing to that one name" issue ...

File: src/main/java/isaac/bastion/CommonSettings.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+public class CommonSettings {
+	private boolean cancelReinInBastionField;
+
+	public boolean isCancelReinInBastionField() {
+		return this.cancelReinInBastionField;
+	}
+
+	public static CommonSettings load(ConfigurationSection config) {
+		CommonSettings settings = new CommonSettings();
+		settings.cancelReinInBastionField = config.getBoolean("cancelReinInBastionField", false);
```
Comment (ProgrammerDan) [inline src/main/java/isaac/bastion/CommonSettings.java]: Might be worth verbosity here: "cancelReinforcementModeInBastionField" vs. shortened form

File: src/main/java/isaac/bastion/listeners/BastionInteractListener.java [hunk 3]
```diff
@@ -145,13 +146,25 @@ public void onBlockPlace(BlockPlaceEvent event) {
...
 	public void onReinforcement(ReinforcementCreationEvent event) {
-		final BastionType type = storage.getAndRemovePendingBastion(event.getBlock().getLocation());
+		if(Bastion.getCommonSettings().isCancelReinInBastionField()) {
+			Set<Block> blocks = new CopyOnWriteArraySet<Block>();
+			blocks.add(event.getBlock());
+			Set<BastionBlock> blocking = blockManager.shouldStopBlock(null, blocks,event.getPlayer().getUniqueId());
+
+			if (blocking.size() != 0 && !groupManager.canPlaceBlock(event.getPlayer(), blocking)){
+				event.setCancelled(true);
+				event.getPlayer().sendMessage(ChatColor.RED + "Bastion prevents this operation");
```
Comment (ProgrammerDan) [inline src/main/java/isaac/bastion/listeners/BastionInteractListener.java]: Maybe, "A Bastion prevents you from reinforcing"

File: src/main/java/isaac/bastion/manager/BastionBlockManager.java [hunk 4]
```diff
@@ -143,10 +144,22 @@ public boolean onCooldown(UUID player, BastionType type){
 		}
 		
 		for(Block block: result) {
-			toReturn.addAll(getBlockingBastions(block.getLocation(),accessors));
+			preblocking.addAll(getBlockingBastions(block.getLocation(),accessors));
 		}
-		
-		return toReturn;
+
```
Comment (ProgrammerDan) [inline src/main/java/isaac/bastion/manager/BastionBlockManager.java]: I did not do this initially because I was worried other plugins are using this method, and changing the behavior without changing the documentation is not good. Even now I'm unsure if this is a good change, but I have never done a full survey to ensure that at the least, none of the civ-plugins will be impacted by the change.

Something to consider.

File: src/main/java/isaac/bastion/storage/BastionGroupStorage.java [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+			}
+
+			log.info("Loaded " + this.groups.size() + " bastion groups with " + validCount + " allowed groups.");
+
+			if(invalidCount > 0) {
+				log.info(" Marked to delete " + invalidCount + " records for non-existent groups.");
+			}
+		} catch (SQLException e) {
+			e.printStackTrace();
+			log.log(Level.SEVERE, "Error loading bastion groups from database, shutting down", e);
```
Comment (ProgrammerDan) [inline src/main/java/isaac/bastion/storage/BastionGroupStorage.java]: For something as severe as shutting down the server? I'm not sure that's indicated here, but if it is, please, make the message a bit more ... dramatic. 

Red text perhaps? dramatic "=====" bars to separate top and bottom of report?

**Review by ProgrammerDan — CHANGES_REQUESTED** [review]
A few thoughts and comments, looking for feedback on some.

File: src/main/java/isaac/bastion/BastionGroup.java (hunk 1 above)
Comment (Aleksey-Terzi) [reply src/main/java/isaac/bastion/BastionGroup.java:12]: Support for this is already handled, for example:

https://github.com/DevotedMC/Bastion/pull/38/files#diff-b530412c860d5e17c39b3300041bbcb5R134

https://github.com/DevotedMC/Bastion/pull/38/files#diff-b530412c860d5e17c39b3300041bbcb5R280

But I will do more testing together when I will test my changes in CastleGates to support "allowed groups"
I will test CG when current changes will be accepted and "interface" approved

File: src/main/java/isaac/bastion/CommonSettings.java (hunk 2 above)
Comment (Aleksey-Terzi) [reply src/main/java/isaac/bastion/CommonSettings.java]: Fixed

File: src/main/java/isaac/bastion/listeners/BastionInteractListener.java (hunk 3 above)
Comment (Aleksey-Terzi) [reply src/main/java/isaac/bastion/listeners/BastionInteractListener.java]: Fixed

File: src/main/java/isaac/bastion/manager/BastionBlockManager.java (hunk 4 above)
Comment (Aleksey-Terzi) [reply src/main/java/isaac/bastion/manager/BastionBlockManager.java]: I've returned back old method and renamed new method to shouldStopBlockByBlockingBastion

File: src/main/java/isaac/bastion/storage/BastionGroupStorage.java (hunk 5 above)
Comment (Aleksey-Terzi) [reply src/main/java/isaac/bastion/storage/BastionGroupStorage.java]: Done.
Since I copy-pasted this from BastionBlockStorage - I've made "original" more dramatic too

Comment (ProgrammerDan) [reply src/main/java/isaac/bastion/storage/BastionGroupStorage.java]: <3 You are best

File: src/main/java/isaac/bastion/manager/BastionGroupManager.java [hunk 6]
```diff
@@ -123,9 +123,7 @@ public boolean canPlaceBlock(Player player, Set<BastionBlock> bastionBlocks) {
 		return true;
 	}
 
-	public Group getAllowedGroup(Player player, BastionBlock bastionBlock, PermissionType perm) {
-		GroupManager groupManager = NameAPI.getGroupManager();
-
+	public Group findFirstAllowedGroup(Player player, BastionBlock bastionBlock) {
```
Comment (ProgrammerDan) [inline src/main/java/isaac/bastion/manager/BastionGroupManager.java]: This is a good naming change, I'd meant to point it out but forgot, good.

**Review by ProgrammerDan — APPROVED** [review]
Let's get it merged and on test so we can move on to other plugins that might need updates, good work

**Comment by ProgrammerDan** [comment]
Build is up on test now, migration worked beautifully
